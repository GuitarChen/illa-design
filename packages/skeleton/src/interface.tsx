import { HTMLAttributes, ReactNode } from "react"

export type ImageSize = number | "small" | "medium" | "large"
export type ImageShape = "circle" | "square"

export interface SkeletonImageProps extends HTMLAttributes<HTMLDivElement> {
  shape?: ImageShape
  size?: ImageSize
  animation?: boolean
}

export interface SkeletonTextProps extends HTMLAttributes<HTMLUListElement> {
  rows?: number
  width?: number | string | Array<number | string>
  animation?: boolean
}

export interface SkeletonProps extends HTMLAttributes<HTMLDivElement> {
  animation?: boolean
  visible?: boolean
  image?: SkeletonImageProps | boolean
  text?: SkeletonTextProps | boolean
  children?: ReactNode
}
