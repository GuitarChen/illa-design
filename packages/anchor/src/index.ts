export * from "./anchor"
export * from "./interface"
