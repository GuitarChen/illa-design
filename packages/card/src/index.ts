export * from "./card"
export * from "./interface"
export * from "./card-grid"
export * from "./meta"
