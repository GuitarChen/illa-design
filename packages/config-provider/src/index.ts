export * from "./config-provider"
export * from "./interface"
export * from "./config-provider-context"
export * from "./config"

// locale
export * from "./locale/en-US"
export * from "./locale/zh-CN"
export * from "./locale/def"
