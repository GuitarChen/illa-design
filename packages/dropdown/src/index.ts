export * from "./dropdown"
export * from "./interface"
