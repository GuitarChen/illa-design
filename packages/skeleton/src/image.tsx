import { isArray } from "@illa-design/system"
import { SkeletonImageProps } from "./interface"
import { applyAnimation, applyImageStyle } from "./style"

export function Image(props: SkeletonImageProps) {
  const { shape = "circle", size = "medium", animation, ...restProps } = props

  return (
    <div
      css={[applyImageStyle({ shape, size }), applyAnimation(animation)]}
      {...restProps}
    ></div>
  )
}
