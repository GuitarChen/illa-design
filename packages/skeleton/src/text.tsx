import { isArray } from "@illa-design/system"
import { SkeletonTextProps } from "./interface"
import { applyAnimation, applyLineStyle, textContainerStyle } from "./style"

export function Text(props: SkeletonTextProps) {
  const { rows = 3, width = "80%", animation, ...restProps } = props

  const lines = Array.from({ length: rows }, (_, idx) => {
    let lineWidth

    if (isArray(width)) {
      lineWidth = width[idx]
    } else if (idx === rows - 1) {
      lineWidth = width
    }

    return (
      <li
        key={idx}
        css={[applyLineStyle(lineWidth), applyAnimation(animation)]}
      ></li>
    )
  })

  return (
    <ul css={textContainerStyle} {...restProps}>
      {lines}
    </ul>
  )
}
