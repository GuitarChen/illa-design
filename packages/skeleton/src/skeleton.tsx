import { forwardRef, ReactElement } from "react"
import { isObject } from "@illa-design/system"
import {
  SkeletonProps,
  SkeletonImageProps,
  SkeletonTextProps,
} from "./interface"
import { Text } from "./text"
import { Image } from "./image"
import { skeletonStyle, skeletonImageStyle, skeletonTextStyle } from "./style"

function getProps(props: boolean | Object) {
  return isObject(props) ? props : {}
}

export const Skeleton = forwardRef<HTMLDivElement, SkeletonProps>(
  (props, ref) => {
    const {
      style,
      className,
      animation,
      visible = true,
      image,
      text = true,
      children,
      ...restProps
    } = props

    function renderText() {
      if (!text) {
        return null
      }

      const textProps = getProps(text) as SkeletonTextProps

      return (
        <Text css={skeletonTextStyle} {...textProps} animation={animation} />
      )
    }

    function renderImage() {
      if (!image) {
        return null
      }
      const imageProps = getProps(image) as SkeletonImageProps

      return (
        <Image css={skeletonImageStyle} {...imageProps} animation={animation} />
      )
    }

    function renderSkeleton() {
      return (
        <div
          ref={ref}
          css={skeletonStyle}
          className={className}
          style={style}
          {...restProps}
        >
          {renderImage()}
          {renderText()}
        </div>
      )
    }

    return visible ? renderSkeleton() : (children as ReactElement)
  },
)

Skeleton.displayName = "Skeleton"
