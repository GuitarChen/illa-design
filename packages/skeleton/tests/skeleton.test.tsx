import { Skeleton } from "../src"
import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom"

test("Skeleton should render with 3 rows", () => {
  render(<Skeleton data-testid="skeleton" />)
  expect(screen.getByTestId("skeleton").querySelectorAll("li")).toHaveLength(3)
})

test("Skeleton renders without text ", () => {
  render(<Skeleton data-testid="skeleton" text={false} />)
  expect(screen.getByTestId("skeleton").querySelectorAll("li")).toHaveLength(0)
})

test("Skeleton renders with 10 rows", () => {
  render(<Skeleton data-testid="skeleton" text={{ rows: 10 }} />)
  expect(screen.getByTestId("skeleton").querySelectorAll("li")).toHaveLength(10)
})

test("Skeleton renders with width", () => {
  render(<Skeleton data-testid="skeleton" text={{ width: 100 }} />)
  expect(
    screen.getByTestId("skeleton").querySelector("li:last-child"),
  ).toHaveStyle("width: 100px")
})

test("Skeleton renders with width array", () => {
  render(<Skeleton data-testid="skeleton" text={{ width: [100, 200, 300] }} />)
  expect(screen.getByTestId("skeleton")).toMatchSnapshot()
})

test("Skeleton renders with image", () => {
  render(<Skeleton data-testid="skeleton" image />)
  expect(screen.getByTestId("skeleton")).toMatchSnapshot()
})

test("Skeleton renders with squrare shape image", () => {
  render(<Skeleton data-testid="skeleton" image={{ shape: "square" }} />)
  expect(screen.getByTestId("skeleton")).toMatchSnapshot()
})

test("Skeleton renders with custom size image", () => {
  render(<Skeleton data-testid="skeleton" image={{ size: 100 }} />)
  expect(screen.getByTestId("skeleton")).toMatchSnapshot()
})

test("Content should be visible when skeleton is invisible", () => {
  render(
    <Skeleton data-testid="skeleton" visible={false}>
      hello
    </Skeleton>,
  )
  expect(screen.getByText("hello")).toBeVisible()
})
