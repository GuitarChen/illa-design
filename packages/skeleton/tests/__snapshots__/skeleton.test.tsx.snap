// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton renders with custom size image 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  background-color: #e5e6eb;
  width: 100px;
  height: 100px;
  border-radius: 50%;
  margin-right: 16px;
}

.emotion-2 {
  list-style: none;
  padding: 0;
  margin: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
}

.emotion-3 {
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-3:not(last-child) {
  margin-bottom: 16px;
}

.emotion-5 {
  width: 80%;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-5:not(last-child) {
  margin-bottom: 16px;
}

<div
  class="emotion-0"
  data-testid="skeleton"
>
  <div
    class="emotion-1"
  />
  <ul
    class="emotion-2"
  >
    <li
      class="emotion-3"
    />
    <li
      class="emotion-3"
    />
    <li
      class="emotion-5"
    />
  </ul>
</div>
`;

exports[`Skeleton renders with image 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  background-color: #e5e6eb;
  width: 48px;
  height: 48px;
  border-radius: 50%;
  margin-right: 16px;
}

.emotion-2 {
  list-style: none;
  padding: 0;
  margin: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
}

.emotion-3 {
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-3:not(last-child) {
  margin-bottom: 16px;
}

.emotion-5 {
  width: 80%;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-5:not(last-child) {
  margin-bottom: 16px;
}

<div
  class="emotion-0"
  data-testid="skeleton"
>
  <div
    class="emotion-1"
  />
  <ul
    class="emotion-2"
  >
    <li
      class="emotion-3"
    />
    <li
      class="emotion-3"
    />
    <li
      class="emotion-5"
    />
  </ul>
</div>
`;

exports[`Skeleton renders with squrare shape image 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  background-color: #e5e6eb;
  width: 48px;
  height: 48px;
  margin-right: 16px;
}

.emotion-2 {
  list-style: none;
  padding: 0;
  margin: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
}

.emotion-3 {
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-3:not(last-child) {
  margin-bottom: 16px;
}

.emotion-5 {
  width: 80%;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-5:not(last-child) {
  margin-bottom: 16px;
}

<div
  class="emotion-0"
  data-testid="skeleton"
>
  <div
    class="emotion-1"
  />
  <ul
    class="emotion-2"
  >
    <li
      class="emotion-3"
    />
    <li
      class="emotion-3"
    />
    <li
      class="emotion-5"
    />
  </ul>
</div>
`;

exports[`Skeleton renders with width array 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-1 {
  list-style: none;
  padding: 0;
  margin: 0;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  overflow: hidden;
}

.emotion-2 {
  width: 100px;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-2:not(last-child) {
  margin-bottom: 16px;
}

.emotion-3 {
  width: 200px;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-3:not(last-child) {
  margin-bottom: 16px;
}

.emotion-4 {
  width: 300px;
  height: 16px;
  background-color: #e5e6eb;
}

.emotion-4:not(last-child) {
  margin-bottom: 16px;
}

<div
  class="emotion-0"
  data-testid="skeleton"
>
  <ul
    class="emotion-1"
  >
    <li
      class="emotion-2"
    />
    <li
      class="emotion-3"
    />
    <li
      class="emotion-4"
    />
  </ul>
</div>
`;
